src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -304,11 +310,16 @@ bool ShenandoahAdaptiveHeuristics::should_start_gc() {
                  byte_size_in_proper_unit(rate), proper_unit_for_byte_size(rate),
                  byte_size_in_proper_unit(allocation_headroom), proper_unit_for_byte_size(allocation_headroom),
                  _spike_threshold_sd);
-    _last_trigger = SPIKE;
+    accept_trigger_with_type(SPIKE);
     return true;
   }
 
-  return ShenandoahHeuristics::should_start_gc();
+  if (ShenandoahHeuristics::should_start_gc()) {
+    _start_gc_is_pending = true;
```

Comment:
I assume there's no race here, i.e., only one thread reads/writes _start_gc_is_pending. If there's a race, make sure it's benign. In either case, _start_gc_is_pending is made "sticky" by this code.

Reply:
There is no race.  A single control thread queries should_start-gc() and that is the same thread that initiates the GC.